### PR TITLE
feat: Make retryPolicy saner in the presence of an expression

### LINKED ISSUE #10980: retryStrategy expression doesn't work for status.phase "Error"
### Pre-requisites

- [X] I have double-checked my configuration
- [X] I can confirm the issues exists when I tested with `:latest`
- [ ] I'd like to contribute the fix myself (see [contributing guide](https://github.com/argoproj/argo-workflows/blob/master/docs/CONTRIBUTING.md))

### What happened/what you expected to happen?

When a workflow node errors and it has a retryStrategy expression looking for lastRetry.status, the retryStrategy is not envoked.

This workflow runs as expected. It is retried 3 times becuse the workflow node fails:
```yaml
apiVersion: argoproj.io/v1alpha1
kind: Workflow
metadata:
  generateName: designed-to-fail-
  labels:
    workflows.argoproj.io/archive-strategy: "false"
spec:
  entrypoint: test
  templates:
  - name: test
    retryStrategy:
      limit: "3"
      expression: lastRetry.status == "Failed"
    container:
      image: ubuntu
      command:
        - sh
        - -c
        - |
          exit 64
```
image

However, if you run the following workflow, wait for the node to be running in the UI and then kill the pod using kubectl. We see that the node is marked as Errored, but the retryStrategy did not kick in, the node was not retried:

```yaml
apiVersion: argoproj.io/v1alpha1
kind: Workflow
metadata:
  generateName: designed-to-error-
  labels:
    workflows.argoproj.io/archive-strategy: "false"
spec:
  entrypoint: test
  templates:
  - name: test
    retryStrategy:
      limit: "3"
      expression: lastRetry.status == "Error"
    container:
      image: ubuntu
      command:
        - sh
        - -c
        - |
          sleep 5000000
```

image

### Version

v3.4.5 > latest

### Paste a small workflow that reproduces the issue. We must be able to run the workflow; don't enter a workflows that uses private images.

```YAML
see above
```

### Logs from the workflow controller

```text
time="2023-04-25T09:45:23.589Z" level=info msg="Processing workflow" namespace=ci workflow=designed-to-error-plflr
time="2023-04-25T09:45:23.590Z" level=info msg="Task-result reconciliation" namespace=ci numObjs=0 workflow=designed-to-error-plflr
time="2023-04-25T09:45:23.590Z" level=info msg="node changed" namespace=ci new.message= new.phase=Running new.progress=0/1 nodeID=designed-to-error-plflr-1299940749 old.message= old.phase=Pending old.progress=0/1 workflow=designed-to-error-plflr
time="2023-04-25T09:45:23.590Z" level=info msg="node designed-to-error-plflr message: retryStrategy.expression evaluated to false" namespace=ci workflow=designed-to-error-plflr
time="2023-04-25T09:45:23.590Z" level=info msg="TaskSet Reconciliation" namespace=ci workflow=designed-to-error-plflr
time="2023-04-25T09:45:23.590Z" level=info msg=reconcileAgentPod namespace=ci workflow=designed-to-error-plflr
time="2023-04-25T09:45:23.590Z" level=info msg="Workflow to be dehydrated" Workflow Size=1862
time="2023-04-25T09:45:23.600Z" level=info msg="Update workflows 200"
time="2023-04-25T09:45:23.600Z" level=info msg="Workflow update successful" namespace=ci phase=Running resourceVersion=64560163 workflow=designed-to-error-plflr
time="2023-04-25T09:45:23.606Z" level=info msg="Create events 201"
time="2023-04-25T09:45:25.036Z" level=info msg="Get leases 200"
time="2023-04-25T09:45:25.042Z" level=info msg="Update leases 200"
time="2023-04-25T09:45:30.050Z" level=info msg="Get leases 200"
time="2023-04-25T09:45:30.054Z" level=info msg="Update leases 200"
time="2023-04-25T09:45:30.470Z" level=info msg="Enforcing history limit for 'git-backup'" namespace=ci workflow=git-backup
time="2023-04-25T09:45:30.470Z" level=info msg="Enforcing history limit for 'renovate-application-definitions'" namespace=ci workflow=renovate-application-definitions
time="2023-04-25T09:45:30.470Z" level=info msg="Enforcing history limit for 'renovate-cloud-infra'" namespace=ci workflow=renovate-cloud-infra
time="2023-04-25T09:45:30.470Z" level=info msg="Enforcing history limit for 'tf-nightly'" namespace=ci workflow=tf-nightly
time="2023-04-25T09:45:33.600Z" level=info msg="Processing workflow" namespace=ci workflow=designed-to-error-plflr
time="2023-04-25T09:45:33.601Z" level=info msg="Task-result reconciliation" namespace=ci numObjs=0 workflow=designed-to-error-plflr
time="2023-04-25T09:45:33.601Z" level=info msg="Workflow pod is missing" namespace=ci nodeName="designed-to-error-plflr(0)" nodePhase=Running recentlyStarted=false workflow=designed-to-error-plflr
time="2023-04-25T09:45:33.601Z" level=info msg="node designed-to-error-plflr-1299940749 phase Running -> Error" namespace=ci workflow=designed-to-error-plflr
time="2023-04-25T09:45:33.601Z" level=info msg="node designed-to-error-plflr-1299940749 message: pod deleted" namespace=ci workflow=designed-to-error-plflr
time="2023-04-25T09:45:33.601Z" level=info msg="node designed-to-error-plflr-1299940749 finished: 2023-04-25 09:45:33.601482457 +0000 UTC" namespace=ci workflow=designed-to-error-plflr
time="2023-04-25T09:45:33.601Z" level=info msg="Node not set to be retried after status: Error" namespace=ci workflow=designed-to-error-plflr
time="2023-04-25T09:45:33.601Z" level=info msg="node designed-to-error-plflr phase Running -> Error" namespace=ci workflow=designed-to-error-plflr
time="2023-04-25T09:45:33.601Z" level=info msg="node designed-to-error-plflr message: pod deleted" namespace=ci workflow=designed-to-error-plflr
time="2023-04-25T09:45:33.601Z" level=info msg="node designed-to-error-plflr finished: 2023-04-25 09:45:33.601633504 +0000 UTC" namespace=ci workflow=designed-to-error-plflr
time="2023-04-25T09:45:33.601Z" level=info msg="TaskSet Reconciliation" namespace=ci workflow=designed-to-error-plflr
time="2023-04-25T09:45:33.601Z" level=info msg=reconcileAgentPod namespace=ci workflow=designed-to-error-plflr
time="2023-04-25T09:45:33.601Z" level=info msg="Updated phase Running -> Error" namespace=ci workflow=designed-to-error-plflr
time="2023-04-25T09:45:33.601Z" level=info msg="Updated message  -> pod deleted" namespace=ci workflow=designed-to-error-plflr
time="2023-04-25T09:45:33.601Z" level=info msg="Marking workflow completed" namespace=ci workflow=designed-to-error-plflr
time="2023-04-25T09:45:33.601Z" level=info msg="Checking daemoned children of " namespace=ci workflow=designed-to-error-plflr
time="2023-04-25T09:45:33.601Z" level=info msg="Workflow to be dehydrated" Workflow Size=1931
time="2023-04-25T09:45:33.607Z" level=info msg="cleaning up pod" action=deletePod key=ci/designed-to-error-plflr-1340600742-agent/deletePod
time="2023-04-25T09:45:33.610Z" level=info msg="Update workflows 200"
time="2023-04-25T09:45:33.610Z" level=info msg="Workflow update successful" namespace=ci phase=Error resourceVersion=64560223 workflow=designed-to-error-plflr
time="2023-04-25T09:45:33.610Z" level=info msg="Create events 201"
time="2023-04-25T09:45:33.611Z" level=info msg="Queueing Error workflow ci/designed-to-error-plflr for delete in 1h0m0s due to TTL"
time="2023-04-25T09:45:33.612Z" level=info msg="Delete pods 404"
time="2023-04-25T09:45:33.614Z" level=info msg="Create events 201"
time="2023-04-25T09:45:33.626Z" level=info msg="Create events 201"
time="2023-04-25T09:45:33.631Z" level=info msg="DeleteCollection workflowtaskresults 200"
time="2023-04-25T09:45:35.062Z" level=info msg="Get leases 200"
time="2023-04-25T09:45:35.067Z" level=info msg="Update leases 200"
```

### Logs from in your workflow's wait container

```text
n/a - it has to be deleted to trigger the error.
```

### PR DESCRIPTION
**Pull request by Joibel**
Most of the explanation for this change is in #10980. When an `expression` is specified in a `retryStrategy` I think it is confusing that the default `retryPolicy` of `OnFailure` still applies. In that event and that only, the default `retryPolicy` becomes `Always` so the `expression` chooses whether or not a retry happens.

If both a `retryPolicy` and `expression`  are provided the previous behavior remains. If no `retryPolicy` and no `expression` are supplied then the default `retryPolicy` of `OnFailure`, the previous behavior, also remains.

Test cases are added to show this works.

I have clarified the current behavior in one commit, so in the event we don't want this change I indent to submit that as a PR.

Fixes #10980

### CODE REVIEW

File: docs/retries.md
Comment (juliev0) [inline docs/retries.md:51]: Great documentation. I didn't know about mermaid for diagramming. Neat. I know the purpose of the diagram is to show on the right what the "default retryPolicy" is. Another option for even more clarity is to have a branch for whether "retryPolicy" is specified in the Workflow, and for the result to be the net "retryPolicy". But I am also okay the way it is.

File: workflow/controller/operator.go
Comment (juliev0) [inline workflow/controller/operator.go:1019]: maybe we can add a log line here so it's obvious the result?

File: workflow/controller/operator.go
Comment (juliev0) [inline workflow/controller/operator.go:1034]: I noticed some conditional logic on line 2018 that checks the value of `retryStrategy.RetryPolicy`. Does that need to be updated?

File: workflow/controller/operator.go
Comment (Joibel) [reply workflow/controller/operator.go]: Ok, done.

File: workflow/controller/operator.go
Comment (Joibel) [reply workflow/controller/operator.go:1025]: Good catch, I've done that section too.

File: docs/retries.md
Comment (Joibel) [reply docs/retries.md:62]: I've redone this, as I agree it's more helpful. Most people won't care how it works under the hood - they just want to know what will happen when they do whatever it is they do, and hopefully this new chart covers that better.

Comment (juliev0) [reply docs/retries.md:62]: Thank you for doing that! One nitpick is: maybe it would be a little less confusing if the result of `Policy Specified?=No; Expression Specified?=No` were just `OnFailure`?

File: workflow/controller/operator.go
Comment (juliev0) [inline workflow/controller/operator.go:2030]: Good idea to add a method here. Nitpick: can we call RetryPolicyActual() once here so we're not repeating calling the same logic over and over? (slight performance benefit, especially if that method were to get more complicated sometime in the future)

File: pkg/apis/workflow/v1alpha1/workflow_types.go
Comment (Joibel) [reply pkg/apis/workflow/v1alpha1/workflow_types.go:1961]: Fixed

File: workflow/controller/operator.go
Comment (Joibel) [reply workflow/controller/operator.go:2031]: Done

**Review by juliev0 — APPROVED** [review]

**Review by juliev0 — CHANGES_REQUESTED** [review]
(should've waited until tests passed to give approval; will approve again once they pass)

**Comment by Joibel** [comment]
> (should've waited until tests passed to give approval; will approve again once they pass)

@juliev0, the github fairies have finally given this the magic green ticks.

**Review by juliev0 — APPROVED** [review]

**Pull request merged**
